src/openloco/windows/station_list.cpp

```diff
@@ -25,28 +45,715 @@ namespace openloco::ui::windows::station_list
+        // Draw widgets and tabs.
+        window->draw(dpi);
+        draw_tabs(window, dpi);
+
+        // Draw company owner image.
+        auto company = companymgr::get(window->number);
+        auto competitor = objectmgr::get<competitor_object>(company->competitor_id);
+        uint32_t image = gfx::recolour(competitor->images[company->var_19], company->colour.primary);
+        uint16_t x = window->x + window->widgets[widx::company_select].left + 1;
+        uint16_t y = window->y + window->widgets[widx::company_select].top + 1;
+        gfx::draw_image(dpi, x, y, image);
+
+        // TODO: locale-based pluralisation.
+        _common_format_args[0] = window->var_83C == 1 ? string_ids::status_num_stations_singular : string_ids::status_num_stations_plural;
+        _common_format_args[1] = window->var_83C;
+
+        // Draw number of stations.
+        gfx::point_t origin = { (int16_t)(window->x + 4), (int16_t)(window->y + window->height - 12) };
+        gfx::draw_string_494B3F(*dpi, &origin, colour::black, string_ids::white_stringid2, &*_common_format_args);
+    }
+
+    // 0x004917BB
+    static void on_dropdown(ui::window* window, widget_index widgetIndex, int16_t itemIndex)
+    {
+        if (widgetIndex != widx::company_select)
+            return;
+
+        company_id_t companyId = dropdown::getCompanyIdFromSelection(itemIndex);
+
+        // Try to find an open station list for this company.
+        auto companyWindow = WindowManager::bringToFront(WindowType::stationList, companyId);
+        if (companyWindow != nullptr)
+            return;
+
+        // If not, we'll turn this window into a window for the company selected.
+        auto company = companymgr::get(companyId);
+        if (company->name == string_ids::empty)
+            return;
+
+        window->number = companyId;
+        window->owner = companyId;
+        window->var_844 = 0;
+        window->row_count = 0;
+
+        sub_4910E8(window);
+
+        window->var_83C = 0;
+        window->row_hover = -1;
+
+        window->call_on_resize();
+        window->call_prepare_draw();
+        window->init_scroll_widgets();
+        window->invalidate();
+    }
+
+    // 0x004917B0
+    static void on_mouse_down(ui::window* window, widget_index widgetIndex)
+    {
+        if (widgetIndex == widx::company_select)
+            dropdown::populateCompanySelect(window, &window->widgets[widgetIndex]);
+    }
+
+    // 0x00491785
+    static void on_mouse_up(ui::window* window, widget_index widgetIndex)
+    {
+        switch (widgetIndex)
+        {
+            case widx::close_button:
+                WindowManager::close(window);
+                break;
+
+            case tab_all_stations:
+            case tab_rail_stations:
+            case tab_road_stations:
+            case tab_airports:
+            case tab_ship_ports:
+            {
+                if (input::is_tool_active(window->type, window->number))
+                    input::cancel_tool();
```

Comment:
The contain can be removed by using `input:: cancel_tool(window->type, window->number)`

Reply:
This is down to personal preference, I guess, but I really prefer the explicit check followed by the cancellation if applicable.